frontend/catalyst/pennylane_extensions.py

```diff
@@ -2601,6 +2602,78 @@ def getArgsAsJAXArrays(self, flat_args):
         return jnpargs
 
 
+def pure_callback(callback_fn, result_type=None):
+    """Pure callback
+
+    Using `pure_callback` allows a user to run python.
+    `pure_callback`s can be used via a decorator:
+
+    ```python
+    @pure_callback
+    def identity(x) -> int:
+        return x
+    ```
+
+    or through a more functional syntax:
+
+    ```python
+    pure_callback(identity, int)(x)
+    ```
+
+    `pure_callback`s are expected to have a return type which matches
+    the return type of the function being called.
+
+    At the moment, `pure_callback`s should not be used inside gradients.
+    """
+
+    signature = inspect.signature(callback_fn)
+    if result_type is None:
+        result_type = signature.return_annotation
+
+    result_type = tree_map(convert_pytype_to_shaped_array, result_type)
+    if result_type is None:
+        raise TypeError("pure_callback requires a result_type")
+
+    @callback
+    def closure(*args, **kwargs) -> result_type:
+        return callback_fn(*args, **kwargs)
+
+    return closure
+
+
+def io_callback(callback_fn):
+    """IO callback
+
+    Using `io_callback` allows a user to run python.
+    `io_callback`s can be used via a decorator:
```

Comment:
How about the following?

```
    Using `io_callback` allows a user to run a python function with side effects inside an `@qjit` context.
    To mark a function as an `io_callback`, one can use a decorator:
```

I don't like the word "defined" here, because it usually means that the user writes the contents of `io_callback`. E.g.,

```python
def(ine) io_callback:
```

Reply:
Great!